dotCMS/src/main/java/com/dotcms/vanityurl/business/VanityUrlAPIImpl.java

```diff
@@ -54,679 +44,248 @@
-    } // addSecondaryVanityURLCacheCollection.
-
-    private void cleanCurrentVanityUrlsPerSite(final SiteLanguageKey siteLanguageKey) {
-
-        final List<CachedVanityUrl> cachedVanityUrls = this.vanityUrlServices.getCachedVanityUrlList
-                (siteLanguageKey.hostId(), siteLanguageKey.languageId());
+    return cachedVanities;
+  }
+  
 
-        if (null != cachedVanityUrls) {
+  
+  @CloseDBIfOpened
+  @Override
+  public Optional<CachedVanityUrl> resolveVanityUrl(final String url, final Host host, final Language language) {
 
-            for (final CachedVanityUrl cachedVanityUrl : cachedVanityUrls) {
 
-                CacheLocator.getVanityURLCache().remove(new CacheVanityKey(
-                        cachedVanityUrl.getSiteId(),
-                        cachedVanityUrl.getLanguageId(),
-                        cachedVanityUrl.getUrl()
-                ).toString());
-            }
-        }
+    // 404 short circuit
+    final Optional<CachedVanityUrl> shortCircuit = cache.getDirectMapping(url, host, language);
+    if(shortCircuit!=null) {
+        return shortCircuit;
     }
 
-    @CloseDBIfOpened
-    @Override
-    public void validateVanityUrl(final Contentlet contentlet) {
+    // tries specific site, language and url
+    Optional<CachedVanityUrl> matched = load(host, language).parallelStream()
+            .filter(cachedVanityUrl ->
+                    cachedVanityUrl.url.equalsIgnoreCase(url)).findAny();
 
-        final User user = getUser(contentlet);
-        Language language =
-                APILocator.getLanguageAPI().getLanguage(user.getLanguageId());
+    // tries specific site, language and pattern
+    if(!matched.isPresent()) {
+      matched = load(host, language).parallelStream()
+              .filter(cachedVanityUrl ->
+                      cachedVanityUrl.pattern.matcher(url).matches()).findAny();
+    }
 
-        language = (null == language ? APILocator.getLanguageAPI().getDefaultLanguage() : language);
+    
+    // try language fallback
+    if (!matched.isPresent()  && !languageAPI.getDefaultLanguage().equals(language) && languageFallback) {
 
-        // check fields
-        checkMissingField(contentlet, language, VanityUrlContentType.ACTION_FIELD_VAR);
-        checkMissingField(contentlet, language, VanityUrlContentType.URI_FIELD_VAR);
+      matched = resolveVanityUrl(url, host, languageAPI.getDefaultLanguage());
+    }
+    
+    // tries SYSTEM_HOST
+    if (!matched.isPresent() && !APILocator.systemHost().equals(host)) {
 
-        final Integer action     =
-                (int)contentlet.getLongProperty(VanityUrlContentType.ACTION_FIELD_VAR);
-        final String uri         =
-                contentlet.getStringProperty(VanityUrlContentType.URI_FIELD_VAR);
+        matched = resolveVanityUrl(url, APILocator.systemHost(), languageAPI.getDefaultLanguage());
```

Comment:
@erickgonzalez, I think this should use the passed in language, not the default language.

Reply:
done